pkg/testutils/testk8sclient.go

```diff
@@ -0,0 +1,80 @@
+package testutils
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	cassdcapi "github.com/k8ssandra/cass-operator/apis/cassandra/v1beta1"
+	k8ssandraapi "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	reaperapi "github.com/k8ssandra/k8ssandra-operator/apis/reaper/v1alpha1"
+	stargateapi "github.com/k8ssandra/k8ssandra-operator/apis/stargate/v1alpha1"
+	promapi "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"github.com/stretchr/testify/assert"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type TestK8sClient struct {
+	client.Client
+	TestState      *testing.T
+	UnsafeGetSync  func(ctx context.Context, key client.ObjectKey, obj client.Object) error
+	UnsafeListSync func(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error
+	timeout        time.Duration
+	tick           time.Duration
+}
+
+// Get IS AN ASYNC SAFE client.Get(), it is not the normal `Get()`! Use UnsafeListSync() if you want regular Get()'s behaviour.
+func (my TestK8sClient) Get(ctx context.Context, key client.ObjectKey, obj client.Object) error {
+	var err error
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() error {
+		defer wg.Done()
+		for {
+			select {
+			case <-time.After(my.tick): // Poll
+				err = my.UnsafeGetSync(ctx, key, obj)
+				if err == nil {
```

Comment:
I investigated the errors in this branch and it turns out most of them are due to the fact that here, you are assuming that the client expects no error. But in some circumstances, an error is perfectly fine, especially a not-found error.

E.g. controller code is full of the following idiom:

```
if err := remoteClient.Get(ctx, key, obj); err == nil {
	// object exists, check if up-to-date
} else if errors.IsNotFound(err) {
	// object does not exist, create it
} else {
	// unexpected error
}
```

With the changes in this branch, the code gets stuck until the async Get call times out.

Reply:
I think this is an interesting comment and might (together with the next one) be behind a lot of my problems here. I know that most of the e2e tests pass, while most of the envtests are failing. 

I'm going to investigate whether this is because the e2e tests use two clients, one to evaluate test conditions, and another within the controllers themselves.

Based on your comment above, I'm thinking that my error was assuming the same held for the envtests, when, in fact, perhaps envtests use a single client to evaluate test conditions which is then also used as the controller's client.